lib/Echidna/Output/Source.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+module Echidna.Output.Source where
+
+import Control.Lens
+import Data.Maybe (fromMaybe, mapMaybe)
+import Data.Text (Text, unlines, pack, unpack)
+import Data.Text.Encoding (decodeUtf8)
+import Data.Text.IO (writeFile)
+import Data.List (nub)
+
+import EVM.Solidity (SourceCache, SrcMap, SolcContract, sourceLines, sourceFiles, runtimeCode, runtimeSrcmap, creationSrcmap)
+import EVM.Debug (srcMapCodePos)
+import Prelude hiding (unlines, writeFile)
+
+import qualified Data.Vector as V
+
+import qualified Data.Map as M
+import qualified Data.Set as S
+
+import Echidna.Exec 
+import Echidna.Types.Tx
+import Echidna.Types.Signature (getBytecodeMetadata)
+
+type FilePathText = Text
+
+saveCoveredCode :: Maybe FilePath -> SourceCache -> [SolcContract] -> CoverageMap -> IO ()
+saveCoveredCode (Just d) sc cs s = let fn = d ++ "/covered.txt"
+                                       cc = ppCoveredCode sc cs s 
+                                   in writeFile fn cc
+saveCoveredCode Nothing  _  _  _ = pure ()
+
+
+-- | Pretty-print the covered code
+ppCoveredCode :: SourceCache -> [SolcContract] -> CoverageMap -> Text
+ppCoveredCode sc cs s | s == mempty = "Coverage map is empty"
+                      | otherwise   =
+  let allLines = M.toList $ sc ^. sourceLines
+  -- ^ Collect all the possible lines from all the files
+      findFile k = fst $ M.findWithDefault ("<no source code>", mempty) k (sc ^. sourceFiles)
+  -- ^ Auxiliary function to get the path for each source file
+      covLines = concatMap (srcMapCov sc s) cs
+  -- ^ List of covered lines during the fuzzing campaing
+  in unlines $ map snd $ concatMap (\(f,vls) ->
+    (mempty, findFile f) :                                                   -- Add a header for each source file to show its complete path
+      filterLines covLines (map ((findFile f,) . decodeUtf8) $ V.toList vls) -- Show the source code for each file with its covered line.
+                                   ) allLines
+
+-- | Filter the lines per file, marking each line
+filterLines :: [Maybe (FilePathText, Int, TxResult)] -> [(FilePathText, Text)] -> [(FilePathText, Text)]
+filterLines []                  ls  = ls
+filterLines (Nothing      : ns) ls  = filterLines ns ls
+filterLines (Just (f,n,r) : ns) ls  = filterLines ns (markLine n r f ls)
+
+-- | Mark one particular line, from a list of lines, keeping the order of them 
+markLine :: Int -> TxResult -> FilePathText -> [(FilePathText, Text)] -> [(FilePathText, Text)] 
+markLine n r cf ls = case splitAt (n-1) ls of
```

Comment:
let's make this multiline and indent a bit less

Reply:
I think now it is better